stan/math/opencl/kernel_cl.hpp

```diff
@@ -82,7 +82,7 @@ class kernel_functor {
                  std::map<const char*, int> options) {
     auto base_opts = opencl_context.base_opts();
     for (auto& it : options) {
-      if (base_opts[it.first] > it.second) {
+      if (base_opts[it.first] != it.second) {
```

Comment:
Can you give us some background on why this change?

Reply:
This is going to override the context calculated value for `base_opts` with the kernel_functor supplied value each time, so we need a better solution here (its why you have failed tests). 

The kernel parameters have to be defined in opencl_context in `base_opts_` and the default value should be overwritten in `opencl_context_base()` if the device info says the default value is not possible. This may seem like you are defining things in 2 places but the idea is:

- define the parameter in `base_opts` and supply it a default value (the default value defined here will only be used if not supplied with the kernel functor or not overwritten by `opencl_context_base()`, most used for enum like defines `UPPER_TO_LOWER`, etc. and if the developer forgets to add a preferred value.)
- if there is a possibility that some device property may restrict its value, overwrite this default value in `opencl_context_base()` (for devices with restrictions (older GPUs and less common architectures))
- supply a prefferred option in the kernel functor

My suggestion would be to change the `LOCAL_SIZE` parameter to something more general like `THREAD_BLOCK_SIZE_1D` and set its default value to a common size useful for any 1D kernel (256 or 512 are usually good magic numbers). This same define can then be used by other kernels with 1D thread grid.
Then have the `opencl_context_base()` overwrite the value it if the max thread block size of the GPU is smaller and
finally supply 64 in the kernel functor if you feel that 64 is the best for this kernel across most GPUs. 


This compile-time kernel parameter stuff is a bit tricky but we havent yet found a way to make this simpler. We need to offer the flexibility of defining your own values and also have a fallback based on GPU properties.

I know that the absence of a thorough doc on this doesnt help, sorry about that. It will be much easier (I hope) once we finish the paper.